sidecar-ffi/src/unix.rs

```diff
@@ -50,22 +59,139 @@ pub extern "C" fn ddog_ph_file_drop(ph: NativeFile) {
     drop(ph)
 }
 
+pub enum AgentRemoteConfigReader {
+    Named(datadog_sidecar::agent_remote_config::AgentRemoteConfigReader<NamedShmHandle>),
+    Unnamed(datadog_sidecar::agent_remote_config::AgentRemoteConfigReader<ShmHandle>),
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_alloc_anon_shm_handle(
+    size: usize,
+    handle: &mut *mut ShmHandle,
+) -> MaybeError {
+    *handle = Box::into_raw(Box::new(try_c!(ShmHandle::new(size))));
+
+    MaybeError::None
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_map_shm(
+    handle: Box<ShmHandle>,
+    mapped: &mut *mut MappedMem<ShmHandle>,
+    pointer: &mut *mut c_void,
+    size: &mut usize,
+) -> MaybeError {
+    let mut memory_mapped = try_c!(handle.map());
+    let slice = memory_mapped.as_slice_mut();
+    *pointer = slice as *mut [u8] as *mut c_void;
+    *size = slice.len();
+
+    *mapped = Box::into_raw(Box::new(memory_mapped));
+
+    MaybeError::None
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_unmap_shm(mapped: Box<MappedMem<ShmHandle>>) -> Box<ShmHandle> {
+    Box::new((*mapped).into())
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_drop_anon_shm_handle(_: Box<ShmHandle>) {}
+
+#[no_mangle]
+pub extern "C" fn ddog_create_agent_remote_config_writer(
+    writer: &mut *mut AgentRemoteConfigWriter<ShmHandle>,
+    handle: &mut *mut ShmHandle,
+) -> MaybeError {
+    let (new_writer, new_handle) = try_c!(datadog_sidecar::agent_remote_config::create_anon_pair());
+    *writer = Box::into_raw(Box::new(new_writer));
+    *handle = Box::into_raw(Box::new(new_handle));
+
+    MaybeError::None
+}
+
 #[no_mangle]
-pub extern "C" fn ddog_sidecar_transport_drop(t: Box<TelemetryTransport>) {
+pub extern "C" fn ddog_agent_remote_config_reader_for_endpoint(
+    endpoint: &Endpoint,
+) -> Box<AgentRemoteConfigReader> {
+    Box::new(AgentRemoteConfigReader::Named(new_reader(endpoint)))
+}
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_agent_remote_config_reader_for_anon_shm(
+    handle: &ShmHandle,
+    reader: &mut *mut AgentRemoteConfigReader,
+) -> MaybeError {
+    *reader = Box::into_raw(Box::new(AgentRemoteConfigReader::Unnamed(try_c!(
+        reader_from_shm(handle.clone())
+    ))));
+
+    MaybeError::None
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_agent_remote_config_write(
+    writer: &AgentRemoteConfigWriter<ShmHandle>,
+    data: ffi::CharSlice,
+) {
+    writer.write(unsafe { data.as_bytes() });
+}
+
+fn ddog_agent_remote_config_read_generic<'a, T>(
+    reader: &'a mut datadog_sidecar::agent_remote_config::AgentRemoteConfigReader<T>,
+    data: &mut ffi::CharSlice<'a>,
+) -> bool
+where
+    T: FileBackedHandle<T> + From<MappedMem<T>>,
+    datadog_sidecar::agent_remote_config::AgentRemoteConfigReader<T>: ReaderOpener<T>,
+{
+    let (new, contents) = reader.read();
+    // c_char may be u8 or i8 depending on target... convert it.
+    let contents: &[c_char] = unsafe { std::mem::transmute(contents) };
```

Comment:
This doesn't work, because on mac it's i8, on linux u8, ... which is the reason for the transmute in the first place, instead of a cast :-X
See comment on the line above.

Reply:
I'm not sure I understand `reader.read` always return a `&[u8]` and it's only the definition of c_char that changes right?